### PR TITLE
Hide `compiler_builtins` in the prelude

### LINKED ISSUE #113533: Macros exported by compiler_builtins should not be in the prelude
compiler_builtins exports some macros for use by its tests, but rustc seems to make these available to any code as if they were part of the prelude. While compiler_builtins should be fixed to not export these, it is also strange that the prelude includes these macros.

```rust
// Compile with: rustc foo.rs --target aarch64-unknown-linux-gnu --crate-type lib
#![no_std]

foreach_cas16!();
```

```
error: unexpected end of macro invocation
 --> src/lib.rs:3:1
  |
3 | foreach_cas16!();
  | ^^^^^^^^^^^^^^^^ missing tokens in macro arguments
  |
note: while trying to match meta-variable `$r#macro:path`
 --> /home/amanieu/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/obj/build/x86_64-unknown-linux-gnu/stage1-std/aarch64-unknown-linux-gnu/release/build/compiler_builtins-0862f16702efa101/out/outlined_atomics.rs:103:51

error: aborting due to previous error
```

### PR DESCRIPTION
**Pull request by Amanieu**
This crate is a private implementation detail. We only need to insert it into the crate graph for linking and should not expose any of its public API.

Fixes #113533

### CODE REVIEW

**Comment by petrochenkov** [comment]
I'd suggest to not generate an `extern crate` item for `compiler_builtins` at all, and link them directly in the same way as profiler builtins or panic runtimes - https://github.com/rust-lang/rust/pull/109421#issuecomment-1478399097.

This also needs a crater run, just in case.

**Comment by petrochenkov** [comment]
Could you add a FIXME saying that https://github.com/rust-lang/rust/pull/113634 is preferable, but we cannot do it right now due to issues?

File: compiler/rustc_builtin_macros/src/standard_library_imports.rs [hunk 1]
```diff
@@ -44,20 +44,37 @@ pub fn inject(
...
-        krate.items.insert(
-            0,
-            cx.item(
-                span,
-                ident,
-                thin_vec![cx.attr_word(sym::macro_use, span)],
-                ast::ItemKind::ExternCrate(None),
-            ),
-        );
+        if name == sym::compiler_builtins {
```
Comment (petrochenkov) [inline compiler/rustc_builtin_macros/src/standard_library_imports.rs:54]:
```rust
let item = if name == sym::compiler_builtins { ... } else { ... };
```
The insertion part (at least) is common.

File: compiler/rustc_builtin_macros/src/standard_library_imports.rs [hunk 2]
```diff
@@ -44,20 +44,37 @@ pub fn inject(
...
+            // compiler_builtins is a private implementation detail. We only
+            // need to insert it into the crate graph for linking and should not
+            // expose any of its public API.
+            krate.items.insert(
+                0,
+                cx.item(
+                    span,
+                    make_ident(kw::Underscore),
```
Comment (petrochenkov) [inline compiler/rustc_builtin_macros/src/standard_library_imports.rs:62]: This will generate `extern crate _ as compiler_builtins;`, no?
How does it work?

**Comment by Amanieu** [comment]
@craterbot check

**Comment by Amanieu** [comment]
Crater failures are all due to running out of disk space.

**Comment by petrochenkov** [comment]
All the regressions are spurious.
@bors r+

**Pull request merged**